File: assignments/ruby/binary-search-tree/binary_search_tree_test.rb
require 'minitest/autorun'
require_relative 'bst'

class BstTest < MiniTest::Unit::TestCase
  def test_data_is_retained
    assert_equal 4, Bst.new(4).data
  end

  def test_inserting_less
    skip
    four = Bst.new 4
    four.insert 2
    assert_equal 4, four.data
    assert_equal 2, four.left.data
  end

  def test_inserting_same
    skip
    four = Bst.new 4
    four.insert 4
    assert_equal 4, four.data
    assert_equal 4, four.left.data
  end

  def test_inserting_right
    skip
    four = Bst.new 4
    four.insert 5
    assert_equal 4, four.data
    assert_equal 5, four.right.data
  end

  def test_complex_tree
    skip
    four = Bst.new 4
    four.insert 2
    four.insert 6
    four.insert 1
    four.insert 3
    four.insert 7
    four.insert 5
    assert_equal 4, four.data
    assert_equal 2, four.left.data
    assert_equal 1, four.left.left.data
    assert_equal 3, four.left.right.data
    assert_equal 6, four.right.data
    assert_equal 5, four.right.left.data
    assert_equal 7, four.right.right.data
  end

  def record_all_data(bst)
    all_data = []
    bst.each { |data| all_data << data }
    all_data
  end

  def test_iterating_one_element
    skip
    assert_equal [4], record_all_data(Bst.new 4)
  end

  def test_iterating_over_smaller_element
    skip
    four = Bst.new 4
    four.insert 2
    assert_equal [2, 4], record_all_data(four)
  end

  def test_iterating_over_larger_element
    skip
    four = Bst.new 4
    four.insert 5
    assert_equal [4, 5], record_all_data(four)
  end

  def test_iterating_over_complex_tree
    skip
    four = Bst.new 4
    four.insert 2
    four.insert 1
    four.insert 3
    four.insert 6
    four.insert 7
    four.insert 5
    assert_equal [1, 2, 3, 4, 5, 6, 7], record_all_data(four)
  end
end

File: assignments/go/binary-search-tree/binary_search_tree_test.go
package binary_search_tree

import (
	"testing"
	"fmt"
)

func TestDataIsRetained(t *testing.T) {
	expected := 4
	actual := Bst(4).data
	if expected != actual {
		t.Errorf("Bst(4).data: expected %d, actual %d", expected, actual)
	}
}

func TestInsertingLess(t *testing.T) {
	bst := SearchTreeData{data: 4}
	bst.Insert(2)

	expected_data := 4
	actual_data := bst.data
	if expected_data != actual_data {
		t.Errorf("bst.data: expected %d, actual %d", expected_data, actual_data)
	}

	expected_left_data := 2
	actual_left_data := bst.left.data
	if expected_left_data != actual_left_data {
		t.Errorf("bst.data: expected %d, actual %d", expected_left_data, actual_left_data)
	}
}

func TestInsertingSame(t *testing.T) {
	bst := SearchTreeData{data: 4}
	bst.Insert(4)

	expected_data := 4
	actual_data := bst.data
	if expected_data != actual_data {
		t.Errorf("bst.data: expected %d, actual %d", expected_data, actual_data)
	}

	expected_left_data := 4
	actual_left_data := bst.left.data
	if expected_left_data != actual_left_data {
		t.Errorf("bst.data: expected %d, actual %d", expected_left_data, actual_left_data)
	}
}

func TestInsertingMore(t *testing.T) {
	bst := SearchTreeData{data: 4}
	bst.Insert(5)

	expected_data := 4
	actual_data := bst.data
	if expected_data != actual_data {
		t.Errorf("bst.data: expected %d, actual %d", expected_data, actual_data)
	}

	expected_right_data := 5
	actual_left_data := bst.right.data
	if expected_right_data != actual_left_data {
		t.Errorf("bst.data: expected %d, actual %d", expected_right_data, actual_left_data)
	}
}

func TestComplexTree(t *testing.T) {
	bst := SearchTreeData{data: 4}
	bst.Insert(2)
	bst.Insert(6)
	bst.Insert(1)
	bst.Insert(3)
	bst.Insert(7)
	bst.Insert(5)

	expected := 4
	actual := bst.data
	if actual != expected {
		t.Errorf("bst.data: expected %d, actual %d", expected, actual)
	}

	expected = 2
	actual = bst.left.data
	if actual != expected {
		t.Errorf("bst.left.data: expected %d, actual %d", expected, actual)
	}

	expected = 1
	actual = bst.left.left.data
	if actual != expected {
		t.Errorf("bst.left.left.data: expected %d, actual %d", expected, actual)
	}

	expected = 3
	actual = bst.left.right.data
	if actual != expected {
		t.Errorf("bst.left.right.data: expected %d, actual %d", expected, actual)
	}

	expected = 6
	actual = bst.right.data
	if actual != expected {
		t.Errorf("bst.right.data: expected %d, actual %d", expected, actual)
	}

	expected = 5
	actual = bst.right.left.data
	if actual != expected {
		t.Errorf("bst.right.left.data: expected %d, actual %d", expected, actual)
	}

	expected = 7
	actual = bst.right.right.data
	if actual != expected {
		t.Errorf("bstfour.right.right.data: expected %d, actual %d", expected, actual)
	}
}

func iToS(i int) string {
	return fmt.Sprintf("%d", i)
}

func TestMapStringWithOneElement(t *testing.T) {
	bst := SearchTreeData{data: 4}

	expected := []string{"4"}
	actual := bst.MapString(iToS)
	if fmt.Sprintf("%s", actual) != fmt.Sprintf("%s", expected) {
		t.Errorf("bst.MapString(): expected %s, actual %s", expected, actual)
	}
}

func TestMapStringWithSmallElement(t *testing.T) {
	bst := SearchTreeData{data: 4}
	bst.Insert(2)

	expected := []string{"2", "4"}
	actual := bst.MapString(iToS)
	if fmt.Sprintf("%v", actual) != fmt.Sprintf("%v", expected) {
		t.Errorf("bst.MapString(): expected %v, actual %v", expected, actual)
	}
}

func TestMapStringWithLargeElement(t *testing.T) {
	bst := SearchTreeData{data: 4}
	bst.Insert(5)

	expected := []string{"4", "5"}
	actual := bst.MapString(iToS)
	if fmt.Sprintf("%v", actual) != fmt.Sprintf("%v", expected) {
		t.Errorf("bst.MapString(): expected %v, actual %v", expected, actual)
	}
}

func TestMapStringWithComplexStructure(t *testing.T) {
	bst := SearchTreeData{data: 4}
	bst.Insert(2)
	bst.Insert(1)
	bst.Insert(3)
	bst.Insert(6)
	bst.Insert(7)
	bst.Insert(5)

	expected := []string{"1", "2", "3", "4", "5", "6", "7"}
	actual := bst.MapString(iToS)
	if fmt.Sprintf("%s", actual) != fmt.Sprintf("%s", expected) {
		t.Errorf("bst.MapString(): expected %s, actual %s", expected, actual)
	}
}

func TestMapIntWithComplexStructure(t *testing.T) {
	bst := SearchTreeData{data: 4}
	bst.Insert(2)
	bst.Insert(1)
	bst.Insert(3)
	bst.Insert(6)
	bst.Insert(7)
	bst.Insert(5)

	f := func(i int) int {
		return i
	}

	expected := []int{1, 2, 3, 4, 5, 6, 7}
	actual := bst.MapInt(f)
	if fmt.Sprintf("%s", actual) != fmt.Sprintf("%s", expected) {
		t.Errorf("bst.MapInt(): expected %s, actual %s", expected, actual)
	}
}
